keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,560 @@
+
+To detect which CronJob has met its schedule and need to create Jobs we need to implement a timer component. We shall implement a Heap based timer algorithm. We will introduce a separate queue with the [`DelayingInterface`](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L37) that implements heap based single shot api [`AddAfter`](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L150). Every time we process an entry from this queue, we will add it back to the queue to simulate a periodic timer.
+
+For further reading:
+1. [Reinventing timer wheel](https://lwn.net/Articles/646950/)
+2. [Hashed and hierarchical timer wheel](http://www.cs.columbia.edu/~nahum/w6998/papers/sosp87-timing-wheels.pdf)
+3. [Golang timers in multi-cpu systems](https://github.com/golang/go/commit/76f4fd8a5251b4f63ea14a3c1e2fe2e78eb74f81)
+4. [Go timerwheel](https://github.com/RussellLuo/timingwheel)
+
+### Metrics
+
+We propose to add metrics that could expose the performance and health of the controller including and not limited to:
+- Skew (actualJobCreationTime-expectedJobCreationTime) - [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
+- Job failures - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- Job successes - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- API server errors - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- Job scheduling latency (podCreationTime - jobCreationTime) - [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
+- Queue depth (pending CronJob and Job entries in the Queue) - [Gauge](https://prometheus.io/docs/concepts/metric_types/#gauge)
+
+Queue depth can be surfaced from existing controller framework.
+
+### Add .status.lastSuccessfulTime
+
+[#issue/75674](https://github.com/kubernetes/kubernetes/issues/75674)
+Add `lastSuccessfulTime` to `.status` that tracks the last time the job completed successfully. This will augment the `lastScheduledTime` available in the `.status` in the v1beta1 api. Potential use is in monitoring (e.g. fire an alert if lastSuccessfulTime is more than X ago).
+
+### Add .status.nextScheduleTime
+
+[#issue/78564](https://github.com/kubernetes/kubernetes/issues/78564)
+Add `nextScheduleTime` to `.status` that tracks the next time the job will be scheduled. This may not be accurate with `Forbid` concurrency policy. This only tracks the `Job` creation time and not the actual `Pod` creation time.
+
+| Concurrency policy | notes |
+|:----------|:----------|
+| Allow   | `nextScheduleTime` would be accurate within a margin of controller scheduling jitter |
+| Forbid  |  `nextScheduleTime` would not be accurate if the previous `Job` takes longer than the cron interval time |
+| Replace | `nextScheduleTime` would be accurate within a margin of controller scheduling jitter along with older concurrent `Job` cleanup if applicable. |
+
```

Comment:
What's the purpose of this field? It's an estimate that can be retrieved from `.status.lastScheduleTime` and the cron spec. If it's not always correct, suggest not putting it in the status to avoid confusion. We could easily add this field after the controller is GA.

Reply:
I don't have any strong motivations, moving to future improvements.